### PR TITLE
Refactor query tower

### PR DESCRIPTION
**Pull request by henrikstranneheim**
## Description

Refactor query tower

### Added

- Tests
- tower_case_config fixture

### Changed

- Refactor query tower function and move it to apps/ Tower
- Removed `query_tower` from MockStore

### Fixed

-

### How to prepare for test
- [ ] ssh to hasta (depending on type of change)
- [ ] activate stage: `us`
- [ ] request trailblazer-stage on hasta: `paxa`
- [ ] install on stage:
    ```Shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_trailblazer -t trailblazer -b [THIS-BRANCH-NAME] -a
    ```
- [ ] ssh to clinical-db (depending on type of change)
- [ ] install on stage:
`bash /home/proj/production/servers/resources/clinical-db.scilifelab.se/update-trailblazer-ui-stage.sh [THIS-BRANCH-NAME]`

### How to test
- [ ] login to ...
- [ ] do ...

### Expected test outcome
- [ ] check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review
- [ ] tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

File: trailblazer/apps/tower/api.py [hunk 1]
```diff
@@ -207,3 +213,21 @@ def _get_job(self, task: TowerTask, analysis_id: int) -> dict:
...
             elapsed=int(task.duration / 60),
         )
+
+
+def _is_tower_api_client_requirements_meet(tower_api: TowerAPI) -> bool:
+    """Raises:
+    TowerRequirementsError when failing meeting Tower mandatory requirement"""
+    if not tower_api.tower_client.meets_requirements:
+        raise TowerRequirementsError("Could not initialize Tower API due to missing requirements")
+    return True
```
Comment (fevac) [inline trailblazer/apps/tower/api.py:223]: aha! is this the good way to do it? Taking notes ✍️

File: tests/apps/tower/test_tower_api.py [hunk 2]
```diff
@@ -171,3 +174,34 @@ def test_tower_task_properties(
...
+    mocker.patch.object(ReadFile, "get_content_from_file")
+    ReadFile.get_content_from_file.return_value = {case_id: [tower_id]}
+
+    # GIVEN Tower requirements are meet
+    mocker.patch(
+        "trailblazer.apps.tower.api._is_tower_api_client_requirements_meet", return_value=True
+    )
+
+    # WHEN getting Tower API
+    tower_api: TowerAPI = get_tower_api(config_file="a_config", case_id=case_id)
```
Comment (fevac) [inline tests/apps/tower/test_tower_api.py:191]: should config_file be a fixture?

**Review by fevac — APPROVED** [review]
Nice! 🌟

File: trailblazer/apps/tower/api.py [hunk 3]
```diff
@@ -207,3 +213,21 @@ def _get_job(self, task: TowerTask, analysis_id: int) -> dict:
...
+
+
+def get_tower_api(config_file: str, case_id: str) -> Optional[TowerAPI]:
```
Comment (diitaz93) [inline trailblazer/apps/tower/api.py:226]:
```suggestion
def get_tower_api(config_file_path: str, case_id: str) -> Optional[TowerAPI]:
```

File: trailblazer/apps/tower/api.py [hunk 4]
```diff
@@ -207,3 +213,21 @@ def _get_job(self, task: TowerTask, analysis_id: int) -> dict:
...
+    """Return Tower API. Currently only one tower ID is supported."""
+    workflow_id: int = ReadFile.get_content_from_file(
+        file_format=FileFormat.YAML, file_path=Path(config_file)
```
Comment (diitaz93) [inline trailblazer/apps/tower/api.py:229]:
```suggestion
        file_format=FileFormat.YAML, file_path=Path(config_file_path)
```

File: trailblazer/apps/tower/api.py [hunk 5]
```diff
@@ -207,3 +213,21 @@ def _get_job(self, task: TowerTask, analysis_id: int) -> dict:
...
+    ).get(case_id)[-1]
+    tower_api = TowerAPI(workflow_id=str(workflow_id))
+    if _is_tower_api_client_requirements_meet(tower_api=tower_api):
```
Comment (diitaz93) [inline trailblazer/apps/tower/api.py:232]: I would move the error rasing of function `_is_tower_api_client_requirements_met` here

**Review by diitaz93 — COMMENTED** [review]
Good refactoring, left some comments

File: trailblazer/apps/tower/api.py [hunk 6]
```diff
@@ -77,15 +80,18 @@ def send_request(self, url: str) -> dict:
...
-        if self.tower_api_endpoint is None or self.tower_api_endpoint == "":
-            LOG.info("Error: no endpoint specified for Tower Open API request.")
-            return False
-        if self.tower_access_token is None or self.tower_access_token == "":
-            LOG.info("Error: no access token specified for Tower Open API request.")
-            return False
-        if self.workspace_id is None or self.workspace_id == "":
-            LOG.info("Error: no workspace specified for Tower Open API request.")
-            return False
+        requirement_map: List[Tuple[str, str]] = [
```
[the remark this answers is not in the document]
Comment (henrikstranneheim) [reply trailblazer/apps/tower/api.py:83]: I guess it is mostly a preference thing. However, this way separates the logic from the definition making it easy to extend any new requirements or remove without changing any logic and the if-tree will most likely grow in the previous version impeding readability.

[the remark this answers is not in the document]
Comment (ChrOertlin) [reply trailblazer/apps/tower/api.py:83]: I think this implementation here is easy to read and understand.

File: trailblazer/apps/tower/api.py [hunk 7]
```diff
@@ -77,15 +80,18 @@ def send_request(self, url: str) -> dict:
...
+            (self.tower_api_endpoint, "Error: no endpoint specified for Tower Open API request."),
+            (
+                self.tower_access_token,
+                "Error: no access token specified for Tower Open API request.",
+            ),
+            (self.workspace_id, "Error: no workspace specified for Tower Open API request."),
+        ]
+        for requirement, msg in requirement_map:
```
Comment (ChrOertlin) [inline trailblazer/apps/tower/api.py:91]:
```suggestion
        for requirement, error_msg in requirement_map:
```

File: trailblazer/apps/tower/api.py [hunk 8]
```diff
@@ -77,15 +80,18 @@ def send_request(self, url: str) -> dict:
...
+            if not requirement:
+                LOG.info(msg)
```
Comment (ChrOertlin) [inline trailblazer/apps/tower/api.py:93]:
```suggestion
                LOG.info(error_msg)
```
Maybe `LOG.warning` or `LOG.error`

File: trailblazer/apps/tower/api.py [hunk 9]
```diff
@@ -207,3 +213,21 @@ def _get_job(self, task: TowerTask, analysis_id: int) -> dict:
             started_at=task.start,
             elapsed=int(task.duration / 60),
         )
+
+
+def _is_tower_api_client_requirements_meet(tower_api: TowerAPI) -> bool:
```
Comment (ChrOertlin) [inline trailblazer/apps/tower/api.py:218]:
```suggestion
def _are_tower_api_client_requirements_met(tower_api: TowerAPI) -> bool:
```

File: trailblazer/apps/tower/api.py [hunk 10]
```diff
@@ -207,3 +213,21 @@ def _get_job(self, task: TowerTask, analysis_id: int) -> dict:
...
+
+
+def get_tower_api(config_file: str, case_id: str) -> Optional[TowerAPI]:
+    """Return Tower API. Currently only one tower ID is supported."""
+    workflow_id: int = ReadFile.get_content_from_file(
+        file_format=FileFormat.YAML, file_path=Path(config_file)
+    ).get(case_id)[-1]
+    tower_api = TowerAPI(workflow_id=str(workflow_id))
+    if _is_tower_api_client_requirements_meet(tower_api=tower_api):
+        return tower_api
```
Comment (Vince-janv) [inline trailblazer/apps/tower/api.py:233]: Why is case_id needed here?

**Review by Vince-janv — APPROVED** [review]
A bit confused regarding `get_tower_api`, the rest looks good! 👍

**Comment by henrikstranneheim** [comment]
Merge conflicts, fixed all comments here: https://github.com/Clinical-Genomics/trailblazer/pull/289

**Pull request closed without merging**